[StableMIR] API to retrieve definitions from crates

celinval opened the pull request:
Add functions to retrieve function definitions and static items from all crates (local and external).

For external crates, we're still missing items from trait implementation and primitives.

r? @compiler-errors: Do you know what is the best way to retrieve the associated items for primitives and trait implementations for external crates? Thanks!

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

celinval committed:
[StableMIR] API to retrieve definitions from crates

Add functions to retrieve function definitions and static items from
all crates (local and external).

For external crates, add a query to retrieve the number of defs in a
foreign crate.

compiler-errors commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -79,6 +80,57 @@ impl<'tcx> Tables<'tcx> {
+                                result.extend(func(self, tcx, *associated_item));
+                            }
+                        }
+                        _ => {}
```
Why aren't you just handling `DefKind::Impl` here?

compiler-errors replied:
Then you won't need to iterate through the inherent impls above, and you get all the trait implementations too.

compiler-errors replied:
You should also just be able to push traits and impls onto the queue. They have module children.

celinval replied:
I thought I tried that, but I might've used the wrong implementation. I might've tried the associated functions not module children. Let me try it. Thanks!

celinval replied:
I tried the following:
```rust
        while let Some(def_id) = queue.pop() {
            if visited.insert(def_id) {
                result.extend(func(self, tcx, def_id));
                match tcx.def_kind(def_id) {
                    DefKind::Mod | DefKind::ForeignMod | DefKind::Impl { .. } | DefKind::Trait => {
                        queue.extend(
                            tcx.module_children(def_id)
                                .iter()
                                .filter_map(|item| item.res.opt_def_id()),
                        )
                    }
                    _ => {}
                }
            }
        }
```

But in that case, I can no longer find method implementations, such as `std::option::Option::<T>::is_some`. Trait implementations are still not showing up. What am I doing wrong?

compiler-errors replied:
I will take a look when I am at my computer.... can you push some code to test this?

compiler-errors commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -79,6 +80,57 @@ impl<'tcx> Tables<'tcx> {
+    where
+        F: Fn(&mut Tables<'tcx>, TyCtxt<'tcx>, DefId) -> Option<T>,
+    {
+        let crate_def = rustc_span::def_id::DefId { index: CRATE_DEF_INDEX, krate };
```
krate.as_def_id()

compiler-errors commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -79,6 +80,57 @@ impl<'tcx> Tables<'tcx> {
+        while let Some(next) = queue.pop() {
+            for def_id in tcx.module_children(next).iter().filter_map(|item| item.res.opt_def_id())
+            {
+                if !visited.contains(&def_id) {
```
```suggestion
                if !visited.insert(&def_id) {
```

compiler-errors commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -97,34 +97,25 @@ impl<'tcx> Tables<'tcx> {
+                        tcx.associated_items(def_id).in_definition_order().map(|item| item.def_id),
+                    ),
+                    DefKind::Struct | DefKind::Enum | DefKind::Union => {
+                        queue.extend(tcx.inherent_impls(def_id));
```
Sorry, I can't remember what I've asked about during this review and what I haven't, but why is this necesary? Inherent impls *should* be regular module children of `DefKind::Mod`.

celinval replied:
Rustc crashes if I try to use module_children. I think module_children only works for the local crate, not for the external crates.

compiler-errors commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -97,34 +97,25 @@ impl<'tcx> Tables<'tcx> {
+                        tcx.module_children(def_id).iter().filter_map(|item| item.res.opt_def_id()),
+                    ),
+                    DefKind::Impl { .. } => queue.extend(
+                        tcx.associated_items(def_id).in_definition_order().map(|item| item.def_id),
```
I think I mentioned previously using `module_children` for impls and for traits, but I guess I take it back -- can you also move the `DefKInd::Trait` arm here? Using `tcx.associated_items` will also make sure we catch the *synthetic* def ids that correspond to RPITITs/AFITs.

Also, no need to use `associated_items().map()` -- there's an `associated_item_def_ids` query.

compiler-errors commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -79,6 +80,48 @@ impl<'tcx> Tables<'tcx> {
+        queue.extend_from_slice(tcx.trait_impls_in_crate(krate));
+        let mut visited = HashSet::default();
+        let mut result = vec![];
+        while let Some(def_id) = queue.pop() {
```
Frankly, actually, now that I think of it, I wonder if this whole approach is somewhat roundabout, and instead the right way of doing this would simply be to simply loop over all of the `DefIndex`es of the given crate.

That's basically what [`iter_local_def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/context.rs.html#1865-1888) does. Perhaps we should just expose some sort of `num_foreign_definitions` that takes a `CrateNum` and maps it to the number of definitions... 

maybe you can extract that from the [`CrateMetadata`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/rmeta/decoder/struct.CrateMetadata.html), or maybe it needs to be encoded...

celinval replied:
That would be much easier. Let me give it a try. Thanks

compiler-errors commented:
@rustbot author

celinval commented:
I implemented the suggested flow. So much cleaner! Thank you

I wanted to double check two things:

1. Should expose the number of definitions as a query or as an untracked method? The latter would require extending `CrateStoreDyn`. I noticed the following comment in this trait:

> /// Note that this trait should probably not be expanding today. All new
> /// functionality should be driven through queries instead!

So I implemented as a query for now. Since this code is not used during codegen, it may still be OK to add it to the trait, so let me know if you have any preference. Changing it shouldn't be hard though.

2. Should we implement the query for the local crate or panic? I opt for the first for completeness for now.

@rustbot ready

compiler-errors commented on compiler/rustc_middle/src/hir/mod.rs:
```diff
@@ -233,4 +233,5 @@ pub fn provide(providers: &mut Providers) {
     providers.in_scope_traits_map = |tcx, id| {
         tcx.hir_crate(()).owners[id.def_id].as_owner().map(|owner_info| &owner_info.trait_map)
     };
+    providers.num_def_ids = |tcx, _| tcx.definitions_untracked().num_definitions();
```
Remove the local provider, it should never be used.

compiler-errors commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -1812,6 +1812,14 @@ rustc_queries! {
 
+    /// Gets the number of definitions in this crate.
+    ///
+    /// This allows external tools to iterate over all definitions in a crate.
```
```suggestion
    /// This allows external tools to iterate over all definitions in a foreign crate.
    ///
    /// This should never be used for the local crate, instead use XXX
```

compiler-errors replied:
iter_local_def_ids or whatever it's called

celinval replied:
I'm also renaming this query to `num_extern_def_ids` to make this restriction more explicit.

compiler-errors commented:
Please squash this into one commit. We don't typically need PRs with this much granularity, especially since the latter basically reimplements the approach in the first half of this PR.

bors commented:
:pushpin: Commit 0ce579f6f3a45ed0bab7f2957ea14683e69412f6 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 0ce579f6f3a45ed0bab7f2957ea14683e69412f6 -->
<!-- homu: {"type":"Approved","sha":"0ce579f6f3a45ed0bab7f2957ea14683e69412f6","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
